Upload command always visible on browser

federicobozzini opened the pull request:
<!--
Thank you for your Pull Request. Please provide a description and review
the requirements below.

Contributors guide: https://github.com/theia-ide/theia/blob/master/CONTRIBUTING.md

Note: Security vulnerabilities should not be disclosed on GitHub, through a PR or any
other means. See SECURITY.md at the root of this repository, to learn how to report
vulnerabilities.
-->

#### What it does
<!-- Include relevant issues and describe how they are addressed. -->

This PR makes the `Upload Files...` menu entry always visible on a browser environment. At the moment the menu entry disappears if no folder is selected on the explorer and I think this might quite counter-intuitive for a lot of users as menu entries don't generally disappear if they are unusable but they just get disabled.

#### How to test
<!-- Explain how a reviewer can reproduce a bug, test new functionality or verify performance improvements. -->

Just start Theia and check the menu entry for uploading files and the `File` menu. It should now always be visible.

#### Review checklist

- [x] As an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)

#### Reminder for reviewers

- As a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

vince-fugnitto reviewed (commented):
@federicobozzini I'm not sure this change works as you intended, a selection of a directory is needed in order for the command to be added in the first place and is not always visible (I think this behavior is odd):

_initially_ (perform a `reset workbench layout`)

![Screen Shot 2022-10-12 at 8 17 16 AM](https://user-images.githubusercontent.com/40359487/195340698-19743216-6a3d-4b4c-befd-e664f6399ba4.png)

_after a directory selection_:

![Screen Shot 2022-10-12 at 8 19 13 AM](https://user-images.githubusercontent.com/40359487/195340852-69e0b607-3df7-4675-a68f-02f094a45e28.png)

federicobozzini commented:
@vince-fugnitto My mistake, it should now be fixed.

This weird behavior comes from the fact that the command handler is transformed by the `FileSelection.CommandHandler` (based on `SelectionCommandHandler`) [which always check if a selection is currently active](https://github.com/eclipse-theia/theia/blob/master/packages/core/src/common/selection-command-handler.ts#L36). Since not every component based on `SelectionCommandHandler` might need this behavior I've removed from there.

`SelectionCommandHandler` is currently used by:

- [`FileSelection.CommandHandler`](https://github.com/eclipse-theia/theia/blob/master/packages/filesystem/src/browser/file-selection.ts#L28) which I think doesn't need this behavior. At the moment it's only used by the `FileSystemCommands.UPLOAD` so my assumption is based on that.
- [`ProblemSelection.CommandHandler`](https://github.com/eclipse-theia/theia/blob/master/packages/markers/src/browser/problem/problem-selection.ts#L30) which needs this behavior. I moved the logic from `SelectionCommandHandler` to this.
- [`BulkEditNodeSelection.CommandHandler`](https://github.com/eclipse-theia/theia/blob/master/packages/bulk-edit/src/browser/bulk-edit-tree/bulk-edit-node-selection.ts#L29) which is unused at the moment.

federicobozzini commented on packages/core/src/common/selection-command-handler.ts:
```diff
@@ -33,7 +33,7 @@ export class SelectionCommandHandler<S> implements CommandHandler {
     isVisible(...args: any[]): boolean {
         const selection = this.getSelection(...args);
-        return !!selection && (!this.options.isVisible || (this.options.isVisible as any)(selection as any, ...args));
+        return !this.options.isVisible || (this.options.isVisible as any)(selection as any, ...args);
```
I've pushed a different solution, with no breaking changes.

vince-fugnitto replied:
I'm trying to think about this one a bit more, I'm not sure the proper or cleanest solution would be to force the `selection-command-handler` to suit our needs but maybe rather create a new `handler`, or even omit using the `selection-command-handler` and instead handle things ourselves with the `selectionService`.

federicobozzini replied:
I agree this issue can be fixed in different ways. My basic understanding is that there is no reason why `SelectionCommandHandler` should force the visibility of item to always consider a selection though. I don't think it's completely unreasonable to let this be decided for every specific command handler getting registered.

In any case I'm happy to change this PR if you have a better suggestion.

vince-fugnitto replied:
By definition and name the `SelectionCommandHandler` is based on selection so that makes sense. If we do not care about the selection like in this case then the handler shouldn't be used and we can instead handle it ourselves rather than try to fit it to our needs for this specific use-case.

colin-grant-work commented:
I would recommend simply rewriting the handler for this item not to be a `FileSelection.CommandHandler`. Since we have access to the `SelectionService` locally, it would be possible to achieve similar behavior, with the desired modifications, without modifying the `FileSelection.CommandHandler` class. Though in principle, I agree that the handler likely shouldn't impose its own constraints on the handlers passed in - setting its own conditions before it runs the handler - but only handle the logic of extracting the selection.

federicobozzini commented:
@vince-fugnitto @colin-grant-work I've change this PR to only change the behavior of the upload button. There is some code repetition but I think that's unfortunately inevitable.

federicobozzini commented on packages/filesystem/src/browser/filesystem-frontend-contribution.ts:
```diff
@@ -120,12 +120,23 @@ export class FileSystemFrontendContribution implements FrontendApplicationContri
-            execute: selection => this.upload(selection)
-        }));
+        commands.registerCommand(FileSystemCommands.UPLOAD, {
+            isEnabled: (args: unknown[]) => {
+                const selection = this.getSelection(args);
+                if (!selection) {
+                    return false;
+                }
+                return this.canUpload(selection);
+            },
+            isVisible: () => !environment.electron.is(),
+            execute: (args: unknown[]) => {
+                const selection = this.getSelection(args);
+                if (!selection) {
+                    return;
+                }
+                this.upload(selection);
+            }
```
You are certainly right.

federicobozzini commented on packages/filesystem/src/browser/filesystem-frontend-contribution.ts:
```diff
@@ -147,6 +158,30 @@ export class FileSystemFrontendContribution implements FrontendApplicationContri
         }
     }
 
+    private getSelection(...args: unknown[]): FileSelection | undefined {
+        const givenSelection = args.length && this.toSelection(args[0]);
+        if (givenSelection) {
+            return givenSelection;
+        }
+        const globalSelection = this.getSingleSelection(this.selectionService.selection);
+        return this.getSingleSelection(globalSelection);
+    };
+
+    private getSingleSelection(arg: unknown): FileSelection | undefined {
+        const selection = this.toSelection(arg);
+        if (selection) {
+            return selection;
+        }
+        if (Array.isArray(arg)) {
+            return arg.find(el => this.toSelection(el));
+        }
+        return undefined;
+    };
+
+    private toSelection(arg: unknown): FileSelection | undefined {
+        return FileSelection.is(arg) ? arg : undefined;
+    }
```
Fair enough. It's still very readable :+1:

federicobozzini commented:
@colin-grant-work @vince-fugnitto Can this PR be merged?

vince-fugnitto reviewed (approved):
The change looks good to me :+1: 

- [x] `upload files...` is always visible
- [x] `upload files...` is enabled if a folder is selected
- [x] `upload files...` is disabled if a file is selected
- [x] uploading files works as expected

colin-grant-work commented:
@federicobozzini, it looks like this is failing the Eclipse Contributor Agreement check because of the shared commits - would you mind pushing a squashed version?

msujew commented:
@federicobozzini Do you mind squashing your commits to fix the ECA check?

federicobozzini commented:
I've pushed a squashed version, thanks.
